pkg/api/dashboard_public.go

```diff
@@ -54,43 +57,72 @@ func (hs *HTTPServer) GetPublicDashboardConfig(c *models.ReqContext) response.Re
 
 // sets public dashboard configuration for dashboard
 func (hs *HTTPServer) SavePublicDashboardConfig(c *models.ReqContext) response.Response {
-	pdc := &models.PublicDashboardConfig{}
-	if err := web.Bind(c.Req, pdc); err != nil {
+	pubdash := &models.PublicDashboard{}
+	if err := web.Bind(c.Req, pubdash); err != nil {
 		return response.Error(http.StatusBadRequest, "bad request data", err)
 	}
 
+	// Always set the org id to the current auth session orgId
+	pubdash.OrgId = c.OrgId
+
 	dto := dashboards.SavePublicDashboardConfigDTO{
-		OrgId:                 c.OrgId,
-		DashboardUid:          web.Params(c.Req)[":uid"],
-		PublicDashboardConfig: pdc,
+		OrgId:           c.OrgId,
+		DashboardUid:    web.Params(c.Req)[":uid"],
+		UserId:          c.UserId,
+		PublicDashboard: pubdash,
 	}
 
-	pdc, err := hs.dashboardService.SavePublicDashboardConfig(c.Req.Context(), &dto)
+	pubdash, err := hs.dashboardService.SavePublicDashboardConfig(c.Req.Context(), &dto)
 	if err != nil {
 		return handleDashboardErr(http.StatusInternalServerError, "Failed to save public dashboard configuration", err)
 	}
 
-	return response.JSON(http.StatusOK, pdc)
+	return response.JSON(http.StatusOK, pubdash)
 }
 
 // QueryPublicDashboard returns all results for a given panel on a public dashboard
-// POST /api/public/dashboard/:uid/panels/:panelId/query
+// POST /api/public/dashboard/:accessToken/panels/:panelId/query
 func (hs *HTTPServer) QueryPublicDashboard(c *models.ReqContext) response.Response {
 	panelId, err := strconv.ParseInt(web.Params(c.Req)[":panelId"], 10, 64)
 	if err != nil {
 		return response.Error(http.StatusBadRequest, "invalid panel ID", err)
 	}
 
+	dashboard, err := hs.dashboardService.GetPublicDashboard(c.Req.Context(), web.Params(c.Req)[":accessToken"])
+	if err != nil {
+		return response.Error(http.StatusInternalServerError, "could not fetch dashboard", err)
```

Comment:
This is somewhat of a nitpick, but I think we have some duplication of effort in BuildPublicDashboardMetricRequest and this method QueryPublicDashboard is doing a bit too much. We're doing parameter substitution for time here and in GetPublicDashboard and also doing publicDashboard.isEnabled check inside BuildPublicDashboardMetricRequest. 

I think it might be worthwhile to refactor this method to pass params down into a method on the dashboardService where we can use the underlying datastore methods to handle any retrieving dashboards/publicdashboards exactly as we need them. Then we just bubble a single response with the info we need to query datasources with. This allows us to trim the method down and only have 2 actions performed at the top level.

Additionally, I think it will make it cleaner for us to move everything into our own public dashboards service in the near future.

Reply:
Example refactor (untested): https://github.com/grafana/grafana/pull/51217